CenterForOpenScience/osf.io
# [ENG-2631] Provider specific metadata

## Purpose
EGAP needs a way to surface their EGAP Identifier in such a way that it can be set after a Registration is complete, and we suspect that other providers may want to have similar fields to surface unique data that their institutions track. This ticket enables that functionality on the backend.

## Changes
* Add an `additional_metadata_fields` field to `RegistrationProvider` that describes provider-specific metadata fields for registrations that are to be editable by Moderators
* Add an `additional_metadata` field to `Registration` along with a `provider_specific_metadata` property to only surface and set just those fields specified by the provider
* Add a new `ContributorOrModeratorOrPublic` permission class to allow selective Moderator write access through the RegistrationDetail API
* Update the `RegistrationSerializer` to surface the `provider_specific_metadata` and to make it editable by moderators.

## QA Notes

New tests confirm that moderators cannot edit other writeable fields on the registration (unless they are also admins/write contributors on the registration) and that non-moderators cannot edit the provider_specific metadata.

New tests also confirm that the serializer and model property correctly work to make it so that only the fields described by the provider are returned and editable through the API.

Finally, new tests confirm expected behavior when a provider does not define any metadata or when the registration does not store a value for a field defined by the provider, including when the `additional_metadata` is uninitialized.

## Documentation


## Side Effects



## Ticket

https://openscience.atlassian.net/browse/ENG-2631
